packages/compiler/src/server/client-config-provider.ts

```diff
@@ -0,0 +1,77 @@
+import { Connection } from "vscode-languageserver/node.js";
+import { ServerHost } from "./types.js";
+
+/**
+ * TypeSpec client-side configuration provider
+ * Inspired by VS Code's WorkspaceConfiguration API for extensibility
+ */
+export interface ClientConfigProvider {
+  /**
+   * Get a configuration value by section key
+   * @param section The configuration section (e.g., "lsp.emit", "formatting.enabled")
+   * @param defaultValue Default value if not found
+   * @returns The configuration value or default
+   */
+  get<T>(section: string): T | undefined;
+  get<T>(section: string, defaultValue: T): T;
+
+  /**
+   * Initialize client configuration with connection and host
+   * @param connection Language server connection
+   * @param host Server host instance
+   */
+  initialize(connection: Connection, host: ServerHost): Promise<void>;
+}
+
+export function createClientConfigProvider(): ClientConfigProvider {
+  let configuration: Record<string, any> = {};
+
+  function get<T>(section: string): T | undefined;
+  function get<T>(section: string, defaultValue: T): T;
+  function get<T>(section: string, defaultValue?: T): T | undefined {
+    const keys = section.split(".");
+    let current: any = configuration;
+
+    for (const key of keys) {
+      if (current && typeof current === "object" && key in current) {
+        current = current[key];
+      } else {
+        return defaultValue;
+      }
+    }
+
+    return current !== undefined ? current : defaultValue;
+  }
+
+  function update(configs: Record<string, any>): void {
+    configuration = { ...configs };
+  }
+
+  async function initialize(connection: Connection, host: ServerHost): Promise<void> {
+    try {
+      const configs = await connection.workspace.getConfiguration("typespec");
+      host.log({ level: "debug", message: "VSCode settings loaded", detail: configs });
+
+      update(configs);
+
+      connection.onDidChangeConfiguration(async (params) => {
+        if (params.settings && params.settings.typespec) {
```

Comment:
in the future, we may need to check settings outside of typespec, it's better to make it more general instead of limited to typespec.